File: maths/0812-largest-triangle-area.go
package maths

import "math"

// https://leetcode.com/problems/largest-triangle-area/

/*
Heron's formula:
Area = √(s(s-a)(s-b)(s-c))
s = (a+b+c)/2
a = √(x2-x1)^2 + (y2-y1)^2)
*/

/*
Shoelace formula for a triangle
Area = ½ |(x₁y₂ + x₂y₃ + x₃y₁) - (y₁x₂ + y₂x₃ + y₃x₁)|
*/

func largestTriangleArea(points [][]int) float64 {
	n := len(points)

	var (
		x1, y1, x2, y2, x3, y3 int
		area                   float64
	)

	for i := range n {
		x1, y1 = points[i][0], points[i][1]
		for j := 1; j < n-1; j++ {
			x2, y2 = points[j][0], points[j][1]
			for k := j + 1; k < n; k++ {
				x3, y3 = points[k][0], points[k][1]
				area = max(
					area,
					0.5*math.Abs(float64(x1*y2+x2*y3+x3*y1)-float64(y1*x2+y2*x3+y3*x1)),
				)
			}
		}
	}

	return area
}

File: maths/0812-largest-triangle-area.py
from typing import List
import unittest

# https://leetcode.com/problems/largest-triangle-area/

# python3 -m unittest maths/0812-largest-triangle-area.py

# Heron's formula:
# Area = √(s(s-a)(s-b)(s-c))
# s = (a+b+c)/2
# a = √(x2-x1)^2 + (y2-y1)^2)

# Shoelace formula for a triangle
# Area = ½ |(x₁y₂ + x₂y₃ + x₃y₁) - (y₁x₂ + y₂x₃ + y₃x₁)|


class Solution(unittest.TestCase):
    def largestTriangleArea(self, points: List[List[int]]) -> float:
        n = len(points)
        larea = 0
        for i in range(n - 1):
            x1, y1 = points[i]
            for j in range(i + 1, n - 1):
                x2, y2 = points[j]
                for k in range(j + 1, n):
                    x3, y3 = points[k]
                    area = 0.5 * abs((x1 * y2 + x2 * y3 + x3 * y1) - (y1 * x2 + y2 * x3 + y3 * x1))
                    larea = max(larea, area)
        return larea

    def test(self):
        for points, expected in [
            ([[0, 0], [0, 1], [1, 0], [0, 2], [2, 0]], 2.00000),
            ([[1, 0], [0, 0], [0, 1]], 0.50000),
        ]:
            output = self.largestTriangleArea(points)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
